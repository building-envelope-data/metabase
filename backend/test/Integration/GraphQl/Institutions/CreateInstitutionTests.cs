using System;
using System.IO;
using System.Threading.Tasks;
using FluentAssertions;
using Metabase.GraphQl.Institutions;
using Snapshooter;
using Snapshooter.Xunit;
using Xunit;

namespace Metabase.Tests.Integration.GraphQl.Institutions
{
    [Collection(nameof(Data.Institution))]
    public sealed class CreateInstitutionTests
      : InstitutionIntegrationTests
    {
        [Fact]
        public async Task AnonymousUser_IsAuthenticationError()
        {
            // Act
            var response =
                await UnsuccessfullyQueryGraphQlContentAsString(
                    File.ReadAllText("Integration/GraphQl/Institutions/CreateInstitution.graphql"),
                    variables: PendingInstitutionInput
                ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(response);
        }

        [Fact]
        public async Task AnonymousUser_CannotCreateInstitution()
        {
            // Act
            await UnsuccessfullyQueryGraphQlContentAsString(
                File.ReadAllText("Integration/GraphQl/Institutions/CreateInstitution.graphql"),
                variables: PendingInstitutionInput
                ).ConfigureAwait(false);
            var response = await GetInstitutions().ConfigureAwait(false);
            // Assert
            Snapshot.Match(response);
        }

        [Theory]
        [MemberData(nameof(EnumerateInstitutionInputs))]
        public async Task LoggedInUser_IsSuccess(
            string key,
            CreateInstitutionInput input
        )
        {
            // Arrange
            var userId = await RegisterAndConfirmAndLoginUser().ConfigureAwait(false);
            // Act
            var response = await CreateInstitution(
                input with
                {
                    OwnerIds = new[] { userId }
                }
                ).ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                SnapshotNameExtension.Create(key),
                matchOptions => matchOptions
                .Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.createInstitution.institution.id").Should().NotBeNullOrWhiteSpace()
                 )
                 .Assert(fieldOptions =>
                 fieldOptions.Field<Guid>("data.createInstitution.institution.uuid").Should().NotBe(Guid.Empty)
                 )
                );
        }

        [Theory]
        [MemberData(nameof(EnumerateInstitutionInputs))]
        public async Task LoggedInUser_CreatesInstitution(
            string key,
            CreateInstitutionInput input
        )
        {
            // Arrange
            var userId = await RegisterAndConfirmAndLoginUser().ConfigureAwait(false);
            // Act
            var (institutionId, institutionUuid) = await CreateInstitutionReturningIdAndUuid(
                input with
                {
                    OwnerIds = new[] { userId }
                }
                ).ConfigureAwait(false);
            var response = await GetInstitutions().ConfigureAwait(false);
            // Assert
            Snapshot.Match(
                response,
                SnapshotNameExtension.Create(key),
                matchOptions => matchOptions
                .Assert(fieldOptions =>
                 fieldOptions.Field<string>("data.institutions.edges[*].node.id").Should().Be(institutionId)
                 )
                .Assert(fieldOptions =>
                 fieldOptions.Field<Guid>("data.institutions.edges[*].node.uuid").Should().Be(institutionUuid)
                 )
                );
        }
    }
}