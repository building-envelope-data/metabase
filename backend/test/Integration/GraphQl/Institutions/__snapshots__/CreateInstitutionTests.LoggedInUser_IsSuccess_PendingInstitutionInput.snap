{"data":{"createInstitution":{"institution":{"id":"SW5zdGl0dXRpb24KZzdmOTY5NzdmOTU1NjRiNTU5ZThhZDM4YWM2YTJlYzkw","uuid":"7f96977f-9556-4b55-9e8a-d38ac6a2ec90","name":"Institution A","abbreviation":"I!A","description":"Best institution ever!","websiteLocator":"https://institution-a.com/"},"errors":null}}}
